src/tir/ir/buffer.cc

```diff
@@ -407,22 +515,31 @@ PrimExpr Buffer::access_ptr(int access_mask, DataType ptr_type, int content_lane
 
 Buffer::Buffer(Var data, DataType dtype, Array<PrimExpr> shape, Array<PrimExpr> strides,
                PrimExpr elem_offset, String name, int data_alignment, int offset_factor,
-               BufferType buffer_type, Span span) {
+               BufferType buffer_type, Array<IntImm> axis_separators, Span span) {
   DataType storage_dtype = dtype;
   // specially handle bool
   if (storage_dtype == DataType::Bool()) {
     storage_dtype = DataType::Int(8);
   }
-  ICHECK(IsPointerType(data->type_annotation, storage_dtype))
-      << "Buffer data field expect to have the right pointer type annotation"
-      << " annotation=" << data->type_annotation << ", storage_dtype=" << storage_dtype;
+  // The buffer dtype may differ from the dtype of the underlying
```

Comment:
The storage type consistency is an assumption that was being made before. Not sure if such generalization is OK. 

Ideally the common type sharing should be done with an explicit cast of pointer type, rather than implicit cast here.

Would be great to leave a TODO to revisit here, cc @Lunderberg @vinx13 @junrushao1994

Reply:
Agreed, explicit casts will be better, and should be included in later updates.  (Maybe also handling bool with explicit casts as well, to minimize the special case code for it.)  Added a TODO to make sure that it gets revisited.